### PR TITLE
Fixes #32819 - Add Index page template

### PR DESCRIPTION
**Pull request by Ron-Lavi**
A data-driven React template to create index pages that contain search bar and table easily.

blocked by:
- #8619
- #8656
- #8655

### CODE REVIEW

File: app/controllers/architectures_controller.rb [hunk 1]
```diff
@@ -39,4 +56,55 @@ def destroy
       process_error
     end
   end
+
+  private
+
+  def canEdit?(id)
```
Comment (timogoebel) [inline app/controllers/architectures_controller.rb]:
```suggestion
  def can_edit?(id)
```

This is ruby, after all ;-)

File: app/controllers/architectures_controller.rb [hunk 2]
```diff
@@ -39,4 +56,55 @@ def destroy
...
+
+  def rows(architectures)
+    rows = []
+    architectures.map do |a|
+      name = can_edit?(a) ? { type: 'link', label: a.name, path: edit_architecture_path(:id => a) } : a.name
+      os = a.operatingsystems.map(&:to_label).to_sentence
+      hosts_count = { type: 'link', label: hosts_count(:architecture)[a], path: hosts_path(:search => "architecture = #{a}") }
+      rows << { cells: [name, os, hosts_count], permissions: row_permissions(a) }
+    end
+    rows
```
Comment (ekohl) [inline app/controllers/architectures_controller.rb]: You don't need the `rows` variable. You can write
```ruby
def rows(architectures)
  architectures.map do |architecture|
    # ...
    { cells: [name, os, hosts_count], permissions: row_permissions(a) }
  end
end

File: app/controllers/architectures_controller.rb [hunk 3]
```diff
@@ -1,11 +1,28 @@
...
+    respond_to do |format|
+      format.html do
+        render "index"
+      end
+      format.json do
+        render json: {
+          rows: rows(@architectures),
+          row_actions: row_actions,
+          columns: columns,
+          item_count: @architectures.count,
```
Comment (ekohl) [inline app/controllers/architectures_controller.rb]: How does this work with pagination? Does it not paginate at all or is this count just the paginated number?

File: app/controllers/architectures_controller.rb [hunk 4]
```diff
@@ -39,4 +56,55 @@ def destroy
       process_error
     end
   end
+
+  private
+
+  def can_edit?(id)
+    authorized_for(controller: :architectures, action: :update, id: id)
```
Comment (timogoebel) [inline app/controllers/architectures_controller.rb]: As this is supposed to be a template, I'm wondering if it makes sense to make this generic and move that code to a concern.

```suggestion
    authorized_for(controller: controller_permission, action: model_of_controller.find_permission_name(:update), id: id)
```

Disclaimer: Totally untested, I doubt this works. Just wanted to point you to some places how we do permission handling.

File: app/controllers/architectures_controller.rb [hunk 5]
```diff
@@ -39,4 +56,55 @@ def destroy
...
+  end
+
+  def can_delete?(id)
+    authorized_for(controller: :architectures, action: :destroy, id: id)
+  end
+
+  def row_permissions(a)
```
Comment (timogoebel) [inline app/controllers/architectures_controller.rb]:
```suggestion
  def row_permissions(model_id)
```

**Review by ekohl — COMMENTED** [review]
Have you thought about data types for columns? For example, a number is typically always right aligned. Dates also have their own handling.

File: app/controllers/architectures_controller.rb (hunk 3 above)
Comment (timogoebel) [reply app/controllers/architectures_controller.rb]: This should probably use `total_entries`.

File: app/controllers/architectures_controller.rb (hunk 4 above)
Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: I would also like this approach, so each controller won't need to implement the permissions helper over and over.

File: app/controllers/architectures_controller.rb (hunk 3 above)
Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: exactly, pagination will work only if the sum of `pages` & `per_page` items are lower than the `item_count`
also, if item_count is 0 we should display the empty state component

File: app/controllers/architectures_controller.rb (hunk 2 above)
Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: thanks! any advice on how plugins could extend it in the easiest way?

**Comment by Ron-Lavi** [comment]
@ekohl for cells data types, I would like to create helpers so the clientside will look for an object with a `type`
and could transform it to the correct element, e.g: for links, we can do something like: `js_link(label, path)`
that will return: `{ type: 'link', label:label, path: path }`

also with permissions it can look like: `js_link_if_authorized(model_id ,label, path)`
which results with something like: `can_edit?(model_id) ? { type: 'link', label: label, path: path} : label`

for numbers and dates, we can create more formatters.

File: app/controllers/architectures_controller.rb (hunk 2 above)
Comment (timogoebel) [reply app/controllers/architectures_controller.rb]: Basically, in a plugin you could do something like this:

```ruby
module ForemanWhatever
  module ArchitecturesControllerExtensions
    extend ActiveSupport::Concern

    module Overrides
      def rows(architectures)
         rows = super
         rows << { cells: ... }
         # ...
         rows
      end
    end

   included do
     prepend Overrides
   end
  end
end
```

That's of course not a proper API, but I think we can get away with this for the beginning until we have some experience what kind of an API would make sense.

Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: nice! thanks @timogoebel

**Comment by ekohl** [comment]
I was wonder how to you intend to send it in JSON. With numbers you can detect it in JSON since there's int/float but there's no date type so you'll need to work around it. Host reports are probably a good example.

File: app/controllers/architectures_controller.rb [hunk 6]
```diff
@@ -1,11 +1,28 @@
...
   include Foreman::Controller::AutoCompleteSearch
   include Foreman::Controller::Parameters::Architecture
 
   before_action :find_resource, :only => [:edit, :update, :destroy]
 
   def index
     @architectures = resource_base_search_and_page.includes(:operatingsystems)
+    respond_to do |format|
+      format.html do
+        render "index"
```
Comment (Ron-Lavi) [inline app/controllers/architectures_controller.rb:12]: Just so we don't break the current architectures page

File: app/controllers/architectures_controller.rb [hunk 7]
```diff
@@ -39,4 +56,55 @@ def destroy
...
+  def row_permissions(a)
+    { canEdit: can_edit?(a), canDelete: can_delete?(a) }
+  end
+
+  def columns
+    [
+      {
+        sortKey: 'name',
+        title: _('Name'),
+        width: 15,
```
Comment (ShimShtein) [inline app/controllers/architectures_controller.rb]: I'm a bit concerned with putting display hints in the data model. I would vote for adding those in the view layer.

File: app/controllers/architectures_controller.rb (hunk 3 above)
Comment (timogoebel) [reply app/controllers/architectures_controller.rb]: Note, we have the `react_pagination_props` helper method. I'm wondering if we actually use this method anywhere.

File: app/helpers/layout_helper.rb [hunk 8]
```diff
@@ -45,7 +45,9 @@ def fetch_user
   end
 
   def index_pages
-    []
+    [
+      { path: '/next_architectures', props: { path: architectures_path } },
```
Comment (Ron-Lavi) [inline app/helpers/layout_helper.rb]: the paths that are registered here will be added to React-router when the app starts,
and the props will be passed to the index page template.

File: app/registries/menu/loader.rb [hunk 9]
```diff
@@ -96,9 +96,8 @@ def self.load
 
       Manager.map :labs_menu do |menu|
         menu.sub_menu :lab_features_menu, :caption => N_('Lab Features'), :icon => 'fa fa-flask' do
-          menu.item :host_wizard,
-            :caption => 'Host Wizard',
-            :url => '/host_wizard'
+          menu.item :host_wizard, :caption => 'Host Wizard', :url => '/host_wizard'
+          menu.item :architectures, :caption => 'Architectures', :url => '/next_architectures'
```
Comment (Ron-Lavi) [inline app/registries/menu/loader.rb]: This is just to make the lab page reachable.

File: config/routes.rb [hunk 10]
```diff
@@ -598,6 +598,7 @@
   end
 
   match 'host_wizard' => 'react#index', :via => :get
+  get 'next_architectures' => 'react#index'
```
Comment (ekohl) [inline config/routes.rb]: I like reusing the `experimental` namespace
```suggestion
  get 'experimental/architectures' => 'react#index'
```

File: app/controllers/architectures_controller.rb (hunk 2 above)
Comment (ShimShtein) [reply app/controllers/architectures_controller.rb]: @timogoebel , what about utilizing RABL for this? We will be able to add more templates, hence adding more nodes. Something similar to the work I do with [facets](https://github.com/theforeman/foreman/blob/bd688271fca6065c08beda14dcb8c9f07099a8a8/app/views/api/v2/hosts/main.json.rabl#L58)?

File: app/registries/menu/loader.rb (hunk 9 above)
Comment (ekohl) [reply app/registries/menu/loader.rb]: This was already wrong for `host_wizard`, but it's better to reuse the Rails helpers for paths: https://guides.rubyonrails.org/routing.html#generating-paths-and-urls-from-code

File: app/controllers/architectures_controller.rb (hunk 3 above)
Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: I also don't see it is used, it was part of the Audits page implementation, which seem to fetch its data from the Audits controller now in https://github.com/theforeman/foreman/blob/develop/app/controllers/audits_controller.rb#L10

File: app/registries/menu/loader.rb (hunk 9 above)
Comment (Ron-Lavi) [reply app/registries/menu/loader.rb]: there is the the `get 'next_architectures' => 'react#index'` in routes.rb which I think is what you refer to.

**Comment by timogoebel** [comment]
> I was wonder how to you intend to send it in JSON. With numbers you can detect it in JSON since there's int/float but there's no date type so you'll need to work around it.

Note, that graphql do has support for data types.

File: app/registries/menu/loader.rb (hunk 9 above)
Comment (ekohl) [reply app/registries/menu/loader.rb]: Yes, but you should never use `/next_architectures` (and `/host_wizard` for that matter). That's a bad pattern and helpers should be used.

**Comment by ekohl** [comment]
> Note, that graphql do has support for data types.

This is a discussion we had somewhere. Can we actually avoid the whole specific JSON endpoint and use GraphQL to make it more generic.

**Comment by Ron-Lavi** [comment]
> > Note, that graphql do has support for data types.
> 
> This is a discussion we had somewhere. Can we actually avoid the whole specific JSON endpoint and use GraphQL to make it more generic.

This is something we are still discussing in the UI interest group, adding @xprazak2 as he actively works on it,
is it supported in all resources? how easy would it be to override it from a plugin?

**Review by ShimShtein — COMMENTED** [review]
Generally I like the idea, it looks neat.

For the extensibility, I would probably encapsulate all the root methods into a single object, and make a list of them extenisble, something along the lines:

``` ruby
# provider definition, each plugin can have one:
class BaseArchitectureData
  def cells(architecture)
    { name: architecture.name }
  end

  def column_definitions
    [
       {
           sort_key: 'name',
           title: _('Name')
       },
    ]
  end

  def actions(architecture)
  end

  def global_actions
  end
end

# On plugin registration a plugin would register its provider:
# This should be wrapped in a nice DSL
column_data_providers << BaseArchitectureData.new
```

And later on

``` ruby
format.json do
  render json: {
     # iterate through the plugins and extract relevant providers:
     page_actions: column_data_providers.map { |provider| provider.global_actions }
```

File: app/registries/menu/loader.rb (hunk 9 above)
Comment (Ron-Lavi) [reply app/registries/menu/loader.rb]: oh right, I guess when we switch the actual page and use a real rails path we could use something like `architectures_path`

File: app/controllers/architectures_controller.rb (hunk 2 above)
Comment (timogoebel) [reply app/controllers/architectures_controller.rb]: @ShimShtein: The more I think about this, I'm asking myself if we're reinventing the graphql API.

Comment (timogoebel) [reply app/controllers/architectures_controller.rb]: Hm. I need to think more about this. Isn't this actually the view layer? I get, why you're asking for templates. Shouldn't this be done in react/JSX?

Comment (ShimShtein) [reply app/controllers/architectures_controller.rb]: It's kind of a middle layer: this is the data that is passed to the UI, but it's not dictated by the UI (as in graphql API). 
IMHO, graphql is a bit different approach - it gives more control about which data to get from the server to the UI components. So for graphql enabled components, the whole extensibility would look a bit different. The components (that are supplied by plugins and core) on the UI side would "vote" which fields they need from the server, and on the server side, each plugin will extend the type to supply those fields.

**Comment by ShimShtein** [comment]
@ekohl @laviro , about data types: I would try and avoid adding this kind of information here. I think it should be the responsibility of the view layer: it's the view layer that should decide which component to show, and those components should decide what data they read from the store, and how they are transforming it.

File: app/controllers/architectures_controller.rb (hunk 2 above)
Comment (timogoebel) [reply app/controllers/architectures_controller.rb]: Yep, true. I believe graphql (or our rest api) should supply the data. We just need a layer to define what and how the data should be presented in the UI (do we need a presenter, perhaps?)

Comment (ShimShtein) [reply app/controllers/architectures_controller.rb]: @timogoebel since we have extensible UI, I think presenters will not be that good of a fit. Maybe something more data-driven, something along the lines of "here is a JSON node with data, use whatever component that fits the most to render it".
In Microsoft world, the components were registered per actual class and then you could throw the actual objects to the UI layer and it would be rendered using this components. We could theoretically mimic the same with a special property (sort of `_object_type`).

**Comment by Ron-Lavi** [comment]
> @ekohl @laviro , about data types: I would try and avoid adding this kind of information here. I think it should be the responsibility of the view layer: it's the view layer that should decide which component to show, and those components should decide what data they read from the store, and how they are transforming it.

I would agree if the data was always in the same structure so the UI would know exactly which cell to transform,
although with plugins extending the data, it would become more difficult to handle it on the client side.
I do think we should provide helpers which will make it easier to write the objects, e.g `js_link(label, path)`
that returns `{ type: 'link', label:label, path: path }`

this is good especially for our project where not many developers will touch the UI code, but definitely could play with the controller response, and in this way, I am pretty convinced we could shift many pages across the app to be faster, better looking, and easier to write.

**Comment by Ron-Lavi** [comment]
> I do think we should provide helpers which will make it easier to write the objects, e.g js_link(label, path)

In addition, I am exploring if we could also tell the UI which custom component to mount if a plugin or even core, wants to render a specific React component. could be something like: `cells: [{ mount: 'ConfirmationModal' }]`

File: app/controllers/architectures_controller.rb (hunk 4 above)
Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: I moved the code to the application controller for now, and it did work!
less code for the resource controller to implement, and I am sure there are more we can refactor to make it easier.

File: app/controllers/architectures_controller.rb (hunk 3 above)
Comment (ekohl) [reply app/controllers/architectures_controller.rb]: > exactly, pagination will work only if the sum of `pages` & `per_page` items are lower than the `item_count`

I think you may not have understood what I said. I think that `@architectures` is already paginated so the count is the paginated count. So this is probably correct (if I read @timogoebel's comment correctly):
```suggestion
          item_count: @architectures.total_entries,
```

File: app/controllers/architectures_controller.rb [hunk 11]
```diff
@@ -39,4 +56,52 @@ def destroy
...
+      },
+      {
+        title: _('Hosts'),
+        width: 15,
+      },
+    ]
+  end
+
+  def rows(architectures)
+    architectures.map do |a|
```
Comment (ekohl) [inline app/controllers/architectures_controller.rb]: I'd avoid `a` and use `architecture`. That makes it easier to read. Or `arch`.

File: app/helpers/layout_helper.rb (hunk 8 above)
Comment (ekohl) [reply app/helpers/layout_helper.rb]: This should also use the helper as mentioned [here](https://github.com/theforeman/foreman/pull/8596/files#r653379610).

File: app/controllers/architectures_controller.rb (hunk 3 above)
Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: Alright, thanks for the clarification.. I saw that count is being used all over.. and indeed it returns the total amount of architectures, not just the paginated.

**Comment by MariSvirik** [comment]
@laviro 
Mocked up template (based on PF4 and CRHC guidelines/best practices)
https://marvelapp.com/prototype/8813483/screen/80289005

File: app/controllers/architectures_controller.rb [hunk 12]
```diff
@@ -1,11 +1,27 @@
...
+      format.html do
+        render "index"
+      end
+      format.json do
+        render json: {
+          rows: rows(@architectures),
+          columns: columns,
+          item_count: @architectures.count,
+          header: _("Architectures"),
+          title: _("Foreman Architectures"), # optional, sets the browser tab title
```
Comment (ekohl) [inline app/controllers/architectures_controller.rb]: I don't think Foreman Architectures is right. Today we also just have Architectures and that's the right thing IMHO. That also makes me think: in a controller we already have the resource type. Can't we use `model_of_controller` to automatically determine the header and title? That again would reduce the duplication.

Perhaps @timogoebel already knows the right methods to call for that.

File: app/controllers/architectures_controller.rb [hunk 13]
```diff
@@ -39,4 +55,33 @@ def destroy
       process_error
     end
   end
+
+  private
+
+  def controller_permission
+    'architectures'
+  end
```
Comment (ekohl) [inline app/controllers/architectures_controller.rb]: Why is this needed? We already have a `controller_permission` in the `ApplicationController` which should derive it correctly. Is that broken?

File: app/controllers/architectures_controller.rb [hunk 14]
```diff
@@ -39,4 +55,33 @@ def destroy
...
+
+  def columns
+    [js_sortable_col(_('Name'), 'name'), _('Operating systems'), _('Hosts')]
+  end
+
+  def rows(architectures)
+    architectures.map do |arch|
+      name = js_link_if_can_edit(arch, edit_architecture_path(:id => arch), arch.name)
+      os = arch.operatingsystems.map(&:to_label).to_sentence
+      hosts_count = js_link_to(hosts_count(:architecture)[arch], hosts_path(:search => "architecture = #{arch}"))
```
Comment (ekohl) [inline app/controllers/architectures_controller.rb]: Should `hosts_count(:architecture)` be stored outside of the loop? Otherwise I'm afraid it may calculate the value for every architecture. Even if it comes from a cache, that's still overhead.

File: webpack/assets/javascripts/react_app/components/common/IndexPageTemplate/components/Table/index.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    error: selectError(state, key),
+    status: selectStatus(state, key),
+  };
+};
+
+// map action dispatchers to props
+const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);
+
+// export connected component
+export default connect(mapStateToProps, mapDispatchToProps)(Table);
```
Comment (ofedoren) [inline webpack/assets/javascripts/react_app/components/common/IndexPageTemplate/components/Table/index.js]: Didn't we want to get rid of this approach?..

File: app/registries/menu/loader.rb (hunk 9 above)
Comment (Ron-Lavi) [reply app/registries/menu/loader.rb]: using `experimental_architectures_path` worked on other files,
in `loader.rb` I guess we need to include some Rails helper?
I also tried doing something like `Rails.application.routes.url_helpers.experimental_architectures_path` which didn't work

File: app/controllers/architectures_controller.rb (hunk 12 above)
Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: it was an example of how some other plugin could override it,
+1 for the reuse, although `model_of_controller` won't be translated well if needed?

**Comment by Ron-Lavi** [comment]
using now the confirm modal from #8619

File: app/controllers/architectures_controller.rb (hunk 2 above)
Comment (Ron-Lavi) [reply app/controllers/architectures_controller.rb]: would it make sense to handle the extendability in a follow-up PR? opened [redmine](https://projects.theforeman.org/issues/32950)

@ShimShtein started to help me with that, and we wrote some pseudo-code over a discussion in https://github.com/laviro/foreman/pull/9

The experimental architecture page won't affect the existing one,
and I am planning to do it in small steps so it will be easier also to review

File: webpack/assets/javascripts/react_app/components/common/IndexPageTemplate/index.js [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+    perPage,
+    page,
+    variant: PaginationVariant[variant],
+    onSetPage: (e, pageNo) => dispatch(onTableSetPage(pageNo, path)),
+    onPerPageSelect: (e, perP) => dispatch(onTablePerPageSelect(perP, path)),
+    perPageOptions: getPerPageOptions(perPage, defaultPerPage),
+  });
+
+  return (
+    <PageLayout
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/IndexPageTemplate/index.js:157]: This is a PF3 based component, so this template will create  pf3 and pf4  mixed pages (and due to the fact this is a template, we will increase our dependency over PF3 and makes our technical dept deeper)

On the host details page, I consumed the `<PageSection>` component from PF4, how complicated it would be to create a  PF4 `PageSection` wrapper for this template?

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/IndexPageTemplate/index.js:157]: Thanks, @amirfefer you are right,
I will address that soon in https://projects.theforeman.org/issues/32991
in a parallel PR, as it may affect many existing pages
